File: spew_test.py
"""
  this is a python script for use with hexchat v2.x.

  It is for figuring out how the hook_* stuff really
  works, because the xchatpython.html document
  has some awfully large holes in it.

  You should load this with `/py console` then
  `/py load spew_test.py` because you really
  don't want this to be automatically loaded up.
"""

# -- config, stuff the user will want to mess with

# server messages to hook_server() onto
#  doccos say you can do numerics too, but I don't know if it translates
#  the server messages properly, or if these are just substring matches and
#  thus I have to get the correct lettercase or numerics depending on the
#  server.
SERVER_EVENTS = ['PRIVMSG', 'NOTICE', 'PART', 'KICK']

# xchat print events to hook into
#   "The event names are available in the Edit Event Texts window."
#   except there IS NO SUCH THING.  In Xchat 2.8.8, go to the menubar
#   and look for Settings > Advanced... > Text Events...  First paramter
#   of xchat.hook_print() are the items in the left column, which you
#   have to guess each of their purposes (but its not difficult).
PRINT_EVENTS = [ 'Channel Message', 'Channel Action', 'Your Message', 'Your Action' ]


# -- init, stuff we do only once. No customer-servicable parts beyond here

__module_name__  = "spew_test"[:12]
__module_version__  = "20121108"[:8]
__module_description__  = "Regugitate what is received by hook_* stuff"[:32]
__module_author__  =  "Mozai <moc.iazom@sesom>"

import hexchat
import json

def hook_spew(word, word_eol, userdata):
  """ /analthread co/42546122 or
      /analthread http://boards.4chan.org/co/res/42546122
  """
  spew = { 'word' : word, 'word_eol': word_eol, 'userdata': userdata }
  print("*** " + json.dumps(spew))
  return hexchat.EAT_NONE


# -- main
print("\002Loading %s v%s\002" % (__module_name__, __module_version__))

hexchat.hook_command('spew', hook_spew, help='regugitate parameters')
print("\002commands:\002 /spew [followed by nonsense data]")

for event in PRINT_EVENTS:
  hexchat.hook_print(event, hook_spew, userdata={'event': event})

for event in SERVER_EVENTS:
  hexchat.hook_server(event, hook_spew, userdata={'event': event})

# things I saw:
# { "userdata": {"event": "Your Message"},
#   "word": ["Mozai", "!8ball am I pretty?", "@"],
#   "word_eol": ["Mozai !8ball am I pretty? @", "!8ball am I pretty? @", "@"] }
#
# *** {
#   "word": [":kate!hexafluorid@Fish-18io58.s.time4vps.cloud", "PRIVMSG", "#wetfish", ":not", "years", "at", "least"],
#   "word_eol": [":kate!hexafluorid@Fish-18io58.s.time4vps.cloud PRIVMSG #wetfish :not years at least", "PRIVMSG #wetfish :not years at least", "#wetfish :not years at least", ":not years at least", "years at least", "at least", "least"],
#   "userdata": {"event": "PRIVMSG"} }
# *** {
#   "word": ["kate", "not years at least"],
#   "word_eol": ["kate not years at least", "not years at least"],
#   "userdata": {"event": "Channel Message"} }

File: spew_test.lua
--[[
  this is a lua script for use with hexchat

  it's for figuring out how the hook_* stuff really
  works.  the hexchat lua document is pretty good
  but sometimes I want to watch the pistons fire.

  https://hexchat.readthedocs.io/en/latest/script_lua.html

  bad idea to leave this in ~/.hexchat/addons/
  better to use /lua load spew_test.lua then
  /lua unload spew_test.lua when you're done
]]

--[[ what to watch
  can we also use numeric events?
  and I don't know if lettercase matters
  bad idea to add "Key Press" to PRINT_EVENTS
]]
local _name, _version, _descr =
    'spew_test.lua', '20180911', 'Regurgitate what the hook_* stuff gets'
local SERVER_EVENTS = { 'PRIVMSG', 'NOTICE', 'PART', 'JOIN', 'KICK' }
local PRINT_EVENTS = { 'Channel Message', 'Channel Action', 'Your Message', 'Your Action' }


local dumps
dumps = function(i)
    if type(i) == 'table' then
        local s = '{ '
        for k,v in pairs(i) do
            if type(k) ~= 'number' then
                k = '"'..k..'"'
            end
            s = s .. '['..k..'] = ' .. dumps(v) .. ', '
        end
        return s .. '} '
    else
        return tostring(i)
    end
end


local hook_spew = function(word, eol, ...)
    print("*** " .. dumps(word))
    print("*** " .. dumps(eol))
    print("*** " .. dumps(...))
    return hexchat.EAT_NONE
end


if hexchat then
    print("Loading " .. _name .. "(/spew argle bargle)")
    hexchat.register(_name, _version, _descr)
    hexchat.hook_command('SPEW', hook_spew, 'regugitate parameters', hexchat.PRI_HIGHEST)
    for _,i in pairs(PRINT_EVENTS) do
        hexchat.hook_print(i, hook_spew, hexchat.PRI_HIGHEST)
    end
    for _,i in pairs(SERVER_EVENTS) do
        hexchat.hook_server(i, hook_spew, hexchat.PRI_HIGHEST)
    end
else
    print("This is a hexchat module; it's sensless to use it outside hexchat")
end

--[=[
  What I saw:

  /spew argle bargle =>
      *** { [1] = spew, [2] = argle, [3] = bargle,}
      *** { [1] = spew argle bargle, [2] = argle bargle, [3] = bargle,}
      *** nil

  I said "harble bargle dimple grimble" in a channel where I have ops =>
      *** { [1] = Mozai, [2] = argle bargle dimple grimble, [3] = @, }
      *** nil
      *** nil

  @kate said "do you ever get so full" =>
      *** { [1] = :kate!hex@<redacted>, [2] = PRIVMSG, [3] = #wetfish, [4] = :do, [5] = you, [6] = ever, [7] = get, [8] = so, [9] = full, }
      *** { [1] = :kate!hex@<redacted> PRIVMSG #wetfish :do you ever get so full, [2] = PRIVMSG #wetfish :do you ever get so full, [3] = #wetfish :do you ever get so full, [4] = :do you ever get so full, [5] = you ever get so full, [6] = ever get so full, [7] = get so full, [8] = so full, [9] = full, }
      *** nil
      *** { [1] = kate, [2] = do you ever get so full, [3] = @, }
      *** nil
      *** nil

  %Kitten emoted "Kitten  rubs her face against the door" =>
      *** { [1] = :Kitten!Kitten@<redacted>, [2] = PRIVMSG, [3] = #wetfish, [4] = :ACTION, [5] = rubs, [6] = her, [7] = face, [8] = against, [9] = the, [10] = door, }
      *** { [1] = :Kitten!Kitten@<redacted> PRIVMSG #wetfish :ACTION rubs her face against the door, [2] = PRIVMSG #wetfish :ACTION rubs her face against the door, [3] = #wetfish :ACTION rubs her face against the door, [4] = :ACTION rubs her face against the door, [5] = rubs her face against the door, [6] = her face against the door, [7] = face against the door, [8] = against the door, [9] = the door, [10] = door, }
      *** nil
      *** { [1] = Kitten, [2] = rubs her face against the door, [3] = %, }
      *** nil
      *** nil
]=]
